### DISCUSSION TITLE
Feedback/Insight for new language server/extension

### DISCUSSION THREAD

**Discussion by ejfasting** [microsoft/vscode-discussions#1017] [Extension Development QnA]
Hello! 

I'm starting on a new language extension and I want to get some feedback/insight from others if my fundamentals are in order before i get started. 
The syntax looks similar to asp.net, where you have HTML and script mixed and divided by <% and %>. The script in this case is both vanilla javascript and custom javascript (custom classes, methods and other things). What makes this a little convoluted is that it doesn't HAVE to be a mix of the two languages, so it could be only javascript. So you end up with 2 scenarios: 
1. Its a mix of HTML and JS, in that case the textDocument contains <% and %>. 
2. Its only JS, then it does NOT contain <% and %>. 

My idea is that i create a LSP Server to handle all of the logic and have it run on the browser webworker (web extension?) so that it also supports vscode.dev and Monaco. Would there be any reason to avoid running it on the browser webworker for local installations? I'm a little unsure why not 'everything runs like that' if it supports all different scenarios: [Extension Host](https://code.visualstudio.com/api/advanced-topics/extension-host)
 
In my case the JS is a mix of default/vanilla JS and some custom things added by me. I'm unsure if Language Services or Request Forwarding is the way to go (both are handling embedded languages), so I'm open to recommendations!

If my reasoning is not totally off this would mean i can mix/look at these two projects to get started:
[lsp-web-extension-sample](https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-web-extension-sample)
[lsp-embedded-language-service](https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-embedded-language-service)

Does anyone see any big pitfalls with my reasoning and/or choice of 'technology'? 

I'm open to any kind of feedback!

//E

**Comment by abrilurena** [comment]
Hi, @ejfasting! I hope you're doing great. It's been a while, so I was wondering - do you still need help with this?

**Reply by ejfasting** [reply]
Hello again @abrilurena !

I've looked at how this is solved in the Monaco editor, and happily they have a method for this:
https://microsoft.github.io/monaco-editor/typedoc/interfaces/languages.typescript.LanguageServiceDefaults.html#addExtraLib 

What would the equivalent be in a vscode extension?
It would pretty much solve what i want to do here, as i just want to add some types to the intellisense for typescript :)

**Reply by ejfasting** [reply]
Closing this thread after a lack of response.

**Comment by ejfasting** [comment]
Hello @abrilurena !

Thank you for responding :) 

Yes, it's been a while so I have done a lot of testing to figure it out. 
I came across this project: https://github.com/volarjs , which pretty much simplifies what I'm trying to do. 
The question "Would there be any reason to avoid running it on the browser webworker for local installations?" still is relevant though, as I have not found any clear answer to why NOT run it on the webworker when running locally. From my testing runnin the client on the webworker works flawlessly both locally and in the cloud. 

I have come across another question during development, though, which is how you would go about injecting a custom *.d.ts into the tsserver. To better explain it: 
between my <% and %> there is typescript. And I'm leveraging the built-in tsserver that comes with vscode. I have some custom types that i want to add to intellisense, so that you get that shipped with the extension/LSP Server. 
An alternative would be to create an NPM for it, and have each user install that in their local project, but i want to ship it with my extension instead. 
I've asked the same question in various different channels the last couple of months but there looks to be no 'easy' way of doing it. 

Any input is appreciated :) 

//Eivind

**Reply by ashtonckj** [reply]
Hi @ejfasting! I stumbled across this discussion while working on something very similar.

I'm currently developing a VS Code extension for Classic ASP (.asp) files where JavaScript is embedded alongside the Classic ASP/HTML. I'm trying to implement VS Code's built-in TypeScript language service for the JavaScript parts, and I also want to provide custom .d.ts definitions with the extension so that users get IntelliSense for objects that aren't normally known to TypeScript.

Your question about injecting custom .d.ts files into tsserver is pretty much the problem I'm running into now as well. I'm currently experimenting with getting the definitions loaded through the language service, but I'm still learning the TypeScript/tsserver side of VS Code extensions, so I haven't found a clean solution yet.

I noticed you mentioned about Volar. Did you ever end up finding a way to ship custom type definitions with the extension and make them available to the TypeScript language service without requiring users to install an npm package?

I'm just curious because the architecture of your problem seems surprisingly close to what I'm trying to achieve with Classic ASP.
¯\\\_(ツ)\_/¯

**Reply by ejfasting** [reply]
Hello @ashtonckj

I ended up using Volar and found it very nice to work with. Volar generates virtual documents which are then sent off to the other languageservices so in my case i ended up injecting an import-statement at the top of the virtual document before it was forwarded. Volar handles integration with the API which is nice.

I also think you can achieve something similar with a [global.d.ts](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/global-d-ts.html) if you just need globally available types but I haven't looked at it in a while so unsure if it fits your usercase.

You could also look into skipping Volar and make your own [Content Mapper](https://github.com/microsoft/typescript-go/pull/4712) which will be available in TS 7. Personally I would recommend investigating this path first since Volar is still on the fence what this change means for the project as a whole. They (the developers behind Volar) are currently working on [migrate vuejs/language-tools to content mapper](https://github.com/vuejs/language-tools/pull/6170) but time will tell if they redefine the position and role for volar.js in v3.

Hope this helps :)

//e

**Reply by ashtonckj** [reply]
Thank you for taking your time to reply to me. I have been looking into what you have mentioned since you replied. I learned a lot~

Thanks again :D
